README.md
> ⚠️ Deprecated - This Craft CMS 2.x plugin is no longer supported. 

<img src="https://github.com/bymayo/barcode/raw/master/screenshots/icon.png" width="50">

# Barcode for Craft 2.x

Barcode is a Craft CMS plugin that allows you to generate a barcode image (Or HTML) in your templates using Twig. 

This was built for a project that required barcodes to be generated in Twig and exported via a PDF.

It uses the PHP Barcode Generator (https://github.com/picqer/php-barcode-generator) library and includes all the options from there.

## Install

- Add the `barcode` directory into your `craft/plugins` directory.
- Navigate to Settings -> Plugins and click the "Install" button.

## Templating

There is only one method that you can use - generate. 

```
{{ craft.barcode.generate(
	{
		code: '8745364267452',
		barcodeImageType: 'svg',
		type: 'TYPE_CODE_128'
	}
) }}
```

## Templating Options

<table>
	<tr>
		<td><strong>Name</strong></td>
		<td><strong>Type</strong></td>
		<td><strong>Default Value</strong></td>
		<td><strong>Required</strong></td>
		<td><strong>Description</strong></td>
	</tr>
	<tr>
		<td>code</td>
		<td>string</td>
		<td>null</td>
		<td>Yes</td>
		<td>The code data the barcode should be generated from.</td>
	</tr>
	<tr>
		<td>barcodeImageType</td>
		<td>string</td>
		<td>null</td>
		<td>Yes</td>
		<td>Image type you want to output the barcode as. Accepts svg, png, jpg or html (Must be lowercase)</td>
	</tr>
	<tr>
		<td>type</td>
		<td>string</td>
		<td>null</td>
		<td>Yes</td>
		<td>The type of barcode you want to output, accepts all the 'Accepted types' from https://github.com/picqer/php-barcode-generator#accepted-types.</td>
	</tr>
	<tr>
		<td>color</td>
		<td>string</td>
		<td>000000</td>
		<td>---</td>
		<td>Foreground colour of the barcode, accepts all HEX codes (Without the #)</td>
	</tr>
	<tr>
		<td>width</td>
		<td>integer</td>
		<td>2</td>
		<td>---</td>
		<td>Width is based on the length of the data, with this you can make the barcode bars wider.</td>
	</tr>
	<tr>
		<td>height</td>
		<td>integer</td>
		<td>30</td>
		<td>---</td>
		<td>Height of the barcode in pixels.</td>
	</tr>
</table>

## Credits

Barcode Scanner by ProSymbols from Noun Project (https://thenounproject.com/prosymbols/)


CHANGELOG.md
# Barcode Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## 1.0.0 - 2018-02-23
### Added
- Initial release of Craft CMS 3 compatible version


src/Barcode.php
<?php
/**
 * BarcodeService plugin for Craft CMS 3.x
 *
 * BarcodeService is a Craft CMS plugin that allows you to generate a barcode image (Or HTML) in your templates using Twig.
 *
 * @link      https://bymayo.co.uk
 * @copyright Copyright (c) 2018 ByMayo
 */

namespace bymayo\barcode;

use bymayo\barcode\services\BarcodeService as BarcodeService;
use bymayo\barcode\variables\BarcodeVariable;

use Craft;
use craft\base\Plugin;
use craft\services\Plugins;
use craft\events\PluginEvent;
use craft\web\twig\variables\CraftVariable;

use yii\base\Event;

/**
 * Craft plugins are very much like little applications in and of themselves. We’ve made
 * it as simple as we can, but the training wheels are off. A little prior knowledge is
 * going to be required to write a plugin.
 *
 * For the purposes of the plugin docs, we’re going to assume that you know PHP and SQL,
 * as well as some semi-advanced concepts like object-oriented programming and PHP namespaces.
 *
 * https://craftcms.com/docs/plugins/introduction
 *
 * @author    ByMayo
 * @package   BarcodeService
 * @since     1.0.0
 *
 * @property  BarcodeService $barcode
 */
class Barcode extends Plugin
{
    // Static Properties
    // =========================================================================

    /**
     * Static property that is an instance of this plugin class so that it can be accessed via
     * BarcodeService::$plugin
     *
     * @var Barcode
     */
    public static $plugin;

    // Public Properties
    // =========================================================================

    /**
     * To execute your plugin’s migrations, you’ll need to increase its schema version.
     *
     * @var string
     */
    public $schemaVersion = '1.0.0';

    // Public Methods
    // =========================================================================

    /**
     * Set our $plugin static property to this class so that it can be accessed via
     * BarcodeService::$plugin
     *
     * Called after the plugin class is instantiated; do any one-time initialization
     * here such as hooks and events.
     *
     * If you have a '/vendor/autoload.php' file, it will be loaded for you automatically;
     * you do not need to load it in your init() method.
     *
     */
    public function init()
    {
        parent::init();
        self::$plugin = $this;

        $this->setComponents([
            'barcodeService' => \bymayo\barcode\services\BarcodeService::class,
        ]);

        // Register our variables
        Event::on(
            CraftVariable::class,
            CraftVariable::EVENT_INIT,
            function (Event $event) {
                /** @var CraftVariable $variable */
                $variable = $event->sender;
                $variable->set('barcode', BarcodeVariable::class);
            }
        );

        // Do something after we're installed
        Event::on(
            Plugins::class,
            Plugins::EVENT_AFTER_INSTALL_PLUGIN,
            function (PluginEvent $event) {
                if ($event->plugin === $this) {
                    // We were just installed
                }
            }
        );

/**
 * Logging in Craft involves using one of the following methods:
 *
 * Craft::trace(): record a message to trace how a piece of code runs. This is mainly for development use.
 * Craft::info(): record a message that conveys some useful information.
 * Craft::warning(): record a warning message that indicates something unexpected has happened.
 * Craft::error(): record a fatal error that should be investigated as soon as possible.
 *
 * Unless `devMode` is on, only Craft::warning() & Craft::error() will log to `craft/storage/logs/web.log`
 *
 * It's recommended that you pass in the magic constant `__METHOD__` as the second parameter, which sets
 * the category to the method (prefixed with the fully qualified class name) where the constant appears.
 *
 * To enable the Yii debug toolbar, go to your user account in the AdminCP and check the
 * [] Show the debug toolbar on the front end & [] Show the debug toolbar on the Control Panel
 *
 * http://www.yiiframework.com/doc-2.0/guide-runtime-logging.html
 */
        Craft::info(
            Craft::t(
                'barcode',
                '{name} plugin loaded',
                ['name' => $this->name]
            ),
            __METHOD__
        );
    }

    // Protected Methods
    // =========================================================================

}


barcode/BarcodePlugin.php
<?php
/**
 * Barcode
 *
 * @author    Jason Mayo
 * @twitter   @madebymayo
 * @package   Barcode
 *
 */

namespace Craft;

class BarcodePlugin extends BasePlugin
{

    public function init()
    {	    
		require_once __DIR__ . '/vendor/autoload.php';
	    parent::init();   
    }

    public function getName()
    {
         return Craft::t('Barcode');
    }

    public function getDescription()
    {
        return Craft::t('Generates barcode images using PHP Barcode Generator');
    }

    public function getDocumentationUrl()
    {
        return 'https://github.com/bymayo/barcode/blob/master/README.md';
    }

    public function getReleaseFeedUrl()
    {
        return 'https://raw.githubusercontent.com/bymayo/barcode/master/releases.json';
    }

    public function getVersion()
    {
        return '1.0';
    }

    public function getSchemaVersion()
    {
        return '1.0';
    }

    public function getDeveloper()
    {
        return 'ByMayo';
    }

    public function getDeveloperUrl()
    {
        return 'bymayo.co.uk';
    }

    public function hasCpSection()
    {
        return false;
    }
}

src/services/BarcodeService.php
<?php
/**
 * BarcodeService for Barcode plugin for Craft CMS 3.x
 *
 * Barcode is a Craft CMS plugin that allows you to generate a barcode image (Or HTML) in your templates using Twig.
 *
 * @link      https://bymayo.co.uk
 * @copyright Copyright (c) 2018 ByMayo
 * @since 1.0.0
 */

namespace bymayo\barcode\services;

use bymayo\barcode\Barcode;

use Craft;
use craft\base\Component;


class BarcodeService extends Component
{

    /**
     * Returns the color in RGB format. 
     * @param $type
     * @param $hex
     * @return mixed
     */
    public function hexToRgb(String $hex, String $type)
    {

        // Respect to http://stackoverflow.com/a/15202130/403180
        list($r, $g, $b) = sscanf($hex, "%02x%02x%02x");

        if ($type === 'png' || $type === 'jpg') {
            return array($r,$g,$b);
        }
        else {
            return '#' . $hex;
        }
    }


    /**
     * Filter in the given options and generate an actual barcode to be used in the gener
     * @param \Picqer\Barcode\ instance $generator
     * @param array $options
     * @return mixed
     */
    public function getBarcode($generator, $options)
    {

        // Array the barcode types
        $barcodeType = [
            'TYPE_CODE_39' => $generator::TYPE_CODE_39,
            'TYPE_CODE_39_CHECKSUM' => $generator::TYPE_CODE_39_CHECKSUM,
            'TYPE_CODE_39E' => $generator::TYPE_CODE_39E,
            'TYPE_CODE_39E_CHECKSUM' => $generator::TYPE_CODE_39E_CHECKSUM,
            'TYPE_CODE_93' => $generator::TYPE_CODE_93,
            'TYPE_STANDARD_2_5' => $generator::TYPE_STANDARD_2_5,
            'TYPE_STANDARD_2_5_CHECKSUM' => $generator::TYPE_STANDARD_2_5_CHECKSUM,
            'TYPE_INTERLEAVED_2_5' => $generator::TYPE_INTERLEAVED_2_5,
            'TYPE_INTERLEAVED_2_5_CHECKSUM' => $generator::TYPE_INTERLEAVED_2_5_CHECKSUM,
            'TYPE_CODE_128' => $generator::TYPE_CODE_128,
            'TYPE_CODE_128_A' => $generator::TYPE_CODE_128_A,
            'TYPE_CODE_128_B' => $generator::TYPE_CODE_128_B,
            'TYPE_CODE_128_C' => $generator::TYPE_CODE_128_C,
            'TYPE_EAN_2' => $generator::TYPE_EAN_2,
            'TYPE_EAN_5' => $generator::TYPE_EAN_5,
            'TYPE_EAN_8' => $generator::TYPE_EAN_8,
            'TYPE_EAN_13' => $generator::TYPE_EAN_13,
            'TYPE_UPC_A' => $generator::TYPE_UPC_A,
            'TYPE_UPC_E' => $generator::TYPE_UPC_E,
            'TYPE_MSI' => $generator::TYPE_MSI,
            'TYPE_MSI_CHECKSUM' => $generator::TYPE_MSI_CHECKSUM,
            'TYPE_POSTNET' => $generator::TYPE_POSTNET,
            'TYPE_PLANET' => $generator::TYPE_PLANET,
            'TYPE_RMS4CC' => $generator::TYPE_RMS4CC,
            'TYPE_KIX' => $generator::TYPE_KIX,
            'TYPE_IMB' => $generator::TYPE_IMB,
            'TYPE_CODABAR' => $generator::TYPE_CODABAR,
            'TYPE_CODE_11' => $generator::TYPE_CODE_11,
            'TYPE_PHARMA_CODE' => $generator::TYPE_PHARMA_CODE,
            'TYPE_PHARMA_CODE_TWO_TRACKS' => $generator::TYPE_PHARMA_CODE_TWO_TRACKS
        ];


        return
            $generator->getBarcode(
                $options['code'],
                $barcodeType[$options['type']],
                (array_key_exists('width', $options)) ? $options['width'] : 2,
                (array_key_exists('height', $options)) ? $options['height'] : 30,
                (array_key_exists('color', $options)) ? $this->hexToRgb($options['color'], $options['barcodeImageType']) : $this->hexToRgb('000000', $options['barcodeImageType'])
            );
    }


    /**
     * Filter what type of barcode is wished, create an outline and call the getBarcode method to render the actual barcode. 
     * @param array $options
     */
    public function generate(array $options)
    {

        switch ($options['barcodeImageType']) {
            case 'svg':
                $generator = new \Picqer\Barcode\BarcodeGeneratorSVG();
                echo '<img src="data:image/svg+xml;base64,' . base64_encode($this->getBarcode($generator, $options)) . '">';
                break;
            case 'png':
                $generator = new \Picqer\Barcode\BarcodeGeneratorPNG();
                echo '<img src="data:image/png;base64,' . base64_encode($this->getBarcode($generator, $options)) . '">';
                break;
            case 'jpg':
                $generator = new \Picqer\Barcode\BarcodeGeneratorJPG();
                echo '<img src="data:image/jpg;base64,' . base64_encode($this->getBarcode($generator, $options)) . '">';
                break;
            case 'html':
                $generator = new \Picqer\Barcode\BarcodeGeneratorHTML();
                echo $this->getBarcode($generator, $options);
                break;
        }
    }
}
